tests/policy_as_code_test.go

```diff
@@ -0,0 +1,239 @@
+		pc.Require().NoError(pc.informer.Informer().RemoveEventHandler(reg))
+		close(ch)
+	}
+	return ch, deferFunc
+}
+
+func (pc *PolicyAsCodeSuite) toPolicy(i interface{}) *v1alpha1.SecurityPolicy {
+	policyCR := v1alpha1.SecurityPolicy{}
+	obj := i.(*unstructured.Unstructured)
+	pc.Require().NoError(runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &policyCR))
+	return &policyCR
+}
+
+func (pc *PolicyAsCodeSuite) toUnstructured(i interface{}) *unstructured.Unstructured {
+	m, err := runtime.DefaultUnstructuredConverter.ToUnstructured(i)
+	pc.Require().NoError(err)
+	return &unstructured.Unstructured{Object: m}
+}
+
+func (pc *PolicyAsCodeSuite) checkPolicyIsDeclarative(id string) {
+	pc.Require().EventuallyWithT(func(collect *assert.CollectT) {
+		policy, err := pc.centralClient.GetPolicy(pc.ctx, &v1.ResourceByID{Id: id})
+		if err != nil {
+			collect.Errorf("Failed to get policy: %s", err.Error())
+		}
+		if policy.Source != storage.PolicySource_DECLARATIVE {
+			collect.Errorf("Policy %s was not marked as declarative in Central", id)
+		}
+	}, time.Second*5, time.Millisecond*30)
+}
+
+func (pc *PolicyAsCodeSuite) updateCRandObserveInCentral(k8sPolicy *v1alpha1.SecurityPolicy, id string) {
+	k8sPolicy.Spec.PolicySections[0].PolicyGroups[0].Values[0].Value = "3"
+	_, err := pc.k8sClient.Update(pc.ctx, pc.toUnstructured(k8sPolicy), metav1.UpdateOptions{})
+	pc.Require().NoError(err)
+
+	pc.Require().EventuallyWithT(func(collect *assert.CollectT) {
+		policy, err := pc.centralClient.GetPolicy(pc.ctx, &v1.ResourceByID{Id: id})
+		if err != nil {
+			collect.Errorf("Failed to get policy: %s", err.Error())
+		}
+		criteriaValue := policy.PolicySections[0].PolicyGroups[0].Values[0].Value
+		if criteriaValue != "3" {
+			collect.Errorf("Policy criteria not updated in Central. Expected 3 but got %s", criteriaValue)
+		}
+	}, time.Second*5, time.Millisecond*30)
+}
+
+func (pc *PolicyAsCodeSuite) deleteCRandObserveInCentral(k8sPolicy *v1alpha1.SecurityPolicy, id string) {
+	pc.Require().NoError(pc.k8sClient.Delete(pc.ctx, k8sPolicy.GetName(), metav1.DeleteOptions{}))
+
+	pc.Require().EventuallyWithT(func(collect *assert.CollectT) {
+		_, err := pc.centralClient.GetPolicy(pc.ctx, &v1.ResourceByID{Id: id})
+		if err != nil {
+			statusErr, _ := status.FromError(err)
+			if statusErr.Code() != codes.NotFound {
+				collect.Errorf("Unexpected error when geting policy: %s", err.Error())
+			}
+		} else {
+			collect.Errorf("Successfully fetched policy %s when it should be deleted", id)
+		}
+	}, time.Second*5, time.Millisecond*30, "Policy CR deletion not propogated to Central")
+}
+
+func (pc *PolicyAsCodeSuite) TearDownSuite() {
+	// TODO: Don't double delete
+	for _, policy := range pc.policies {
+		log.Infof("Deleting policy with name \"%s\"", policy.Name)
+		_, err := pc.centralClient.DeletePolicy(pc.ctx, &v1.ResourceByID{
+			Id: policy.Id,
+		})
+		pc.Require().NoError(err)
+	}
+	// TODO: Remove finalizers if delete fails
```

Comment:
Yes, when I was testing the deletion part of the test, when it failed, I would have to go manually remove the finalizer to clean up the test because the controller had a bug where it wouldn't remove it.  In this case, any retry of the test would fail because the policy CR would still exist from the previous run.